### PR TITLE
Fused RMSNorm Implementation

### PR DESCRIPTION
**Pull request by AaronWang04**
Implements #72643 
Rebased from #153666 

cc @ptrblck @msaroufim @eqy @jerryzh168

### CODE REVIEW

File: torch/testing/_internal/common_modules.py [hunk 1]
```diff
@@ -4110,6 +4110,9 @@ def module_error_inputs_torch_nn_Pad3d(module_info, device, dtype, requires_grad
                ),
     ModuleInfo(torch.nn.RMSNorm,
                module_inputs_func=module_inputs_torch_nn_RMSNorm,
+               skips=(
+                   # No channels_last support for RMSNorm currently.
```
Comment (vadimkantorov) [inline torch/testing/_internal/common_modules.py:4114]: I think in practcice, layernorm/rmsnorm are used for channels_last inputs and to the several trailing shape dimensions ("BTC"), right? but the logical layout matches the memory layout as is typically

the distinction might make sense if we allow processing of not trailing shape dims

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 2]
```diff
@@ -5022,6 +5022,103 @@ std::tuple<Tensor, Tensor, Tensor> layer_norm_double_backward(
...
+  for (int i = 0; i < normalized_ndim; ++i) {
+    N_rms *= input_shape[axis + i];
+  }
+
+  Tensor dX;
+  Tensor dgamma;
+
+  std::vector<int64_t> rstd_view_shape = rstd.sizes().vec();
+  for (int i = 0;
+       i < std::max(static_cast<int>(normalized_ndim - rstd.sizes().size()), 0);
```
Comment (Skylion007) [inline torch/csrc/autograd/FunctionsManual.cpp]:
```suggestion
       i < std::max(static_cast<int>(normalized_ndim - rstd.dim()), 0);
```
There is a way to access this wihtout return sizes, I thikn it's the above. use dim() or ndimension()

File: torch/testing/_internal/common_modules.py (hunk 1 above)
Comment (eqy) [reply torch/testing/_internal/common_modules.py:4114]: This is grandfathered in from layernorm, and I believe it refers to 4D/5D (NHWC/NHWDC) inputs
